### PR TITLE
Added Rocket Sledge

### PR DESCRIPTION
**Pull request by Waselon**
## About The Pull Request

Added rocket sledge. Power weapon that can be optained through requisitions for 50 points.
has 2 modes (works only when fueled):
1) stun = 1, weaken = 2, stagger = 2, knockback = 0;
2) stun = 1, weaken = 1, stagger = 1, knockback = 1;

75 damage without fuel;
150 damage with fuel (active);
attack speed = 20 (2 sec delay);

can store 50 fuel inside, uses 5 fuel per hit;
can only be used while wielded;
has bump attack;
can be AltClicked to tighten grip;

изображение
изображение

## Why It's Good For The Game

More melee weapons, first melee weapon obtainable through reqisitions.

## Changelog
:cl:
add: Added rocket sledge
imageadd: added rocket sledge sprites
code: changed welder pack and engieneer satchel to refuel rocket sledge
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
Is this the next melee spec

**Comment by chesse20** [comment]
whys it worse than a free melee when unpowered

**Comment by Waselon** [comment]
> whys it worse than a free melee when unpowered if its 50 points

so there's a reason to keep it powered

**Comment by Sebbe9123** [comment]
Not a fan of it having a stun unpowered.

**Comment by InterroLouis** [comment]
It doesn't have a stun unpowered.

**Comment by Araulius** [comment]
Please God let this get merged

**Review by Kuro020 — CHANGES_REQUESTED** [review]
sprite does not meet quality standards, still needs a lot of work, specially on the shape and shading of the hammer head part

**Comment by Walarks** [comment]
Why does it have bump attack? A hammer shouldn't be as easy to handle and hit stuff with as a lightweight sword.

**Comment by Waselon** [comment]
> Why does it have bump attack? A hammer shouldn't be as easy to handle and hit stuff with as a lightweight sword.

for convenient use, I guess

**Comment by Waselon** [comment]
изображение

**Comment by Waselon** [comment]
изображение

**Comment by InterroLouis** [comment]
> Why does it have bump attack? A hammer shouldn't be as easy to handle and hit stuff with as a lightweight sword.

It also has a 2 second attack delay and NEEDS to be wielded in order to actually be useable, and it needs to constantly be fed fuel.

**Comment by Losenis** [comment]
image

File: code/game/objects/items/storage/backpack.dm [hunk 1]
```diff
@@ -706,6 +706,17 @@
 		to_chat(user, span_notice("You refill [FT] with [lowertext(FT.caliber)]."))
 		FT.update_icon()
 
+	else if(istype(I, /obj/item/weapon/twohanded/rocketsledge))
+		var/obj/item/weapon/twohanded/rocketsledge/FT = I
```
Comment (hyper2snyper) [inline code/game/objects/items/storage/backpack.dm:710]: dont use FT for a rocket sledge. FT means flamer tank.

File: code/game/objects/items/tools/maintenance_tools.dm [hunk 2]
```diff
@@ -397,6 +397,17 @@
 		to_chat(user, span_notice("You refill [FT] with [lowertext(FT.caliber)]."))
 		FT.update_icon()
 
+	else if(istype(I, /obj/item/weapon/twohanded/rocketsledge))
+		var/obj/item/weapon/twohanded/rocketsledge/FT = I
```
Comment (hyper2snyper) [inline code/game/objects/items/tools/maintenance_tools.dm:401]: ditto

File: code/game/objects/items/weapons/twohanded.dm [hunk 3]
```diff
@@ -338,9 +338,162 @@
...
+	force_wielded = 75
+	throwforce = 50
+	throw_speed = 2
+	edge = 1
+	sharp = IS_SHARP_ITEM_BIG
+	flags_atom = CONDUCT | TWOHANDED
+	attack_verb = list("smashed", "hammered")
+	attack_speed = 20
+
+	var/max_fuel = 50 //amount of fuel stored inside
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:364]:
```suggestion
	///amount of fuel stored inside
	var/max_fuel = 50
```
autodoc with three `/`
like `///`
and all the vars need to be autodocd.

File: code/game/objects/items/weapons/twohanded.dm [hunk 4]
```diff
@@ -338,9 +338,162 @@
...
+
+/obj/item/weapon/twohanded/rocketsledge/dropped(mob/user)
+	. = ..()
+	toggle_item_bump_attack(user, FALSE)
+
+/obj/item/weapon/twohanded/rocketsledge/examine(mob/user)
+	..()
+	to_chat(user, "It contains [get_fuel()]/[max_fuel] units of fuel!")
+
+/obj/item/weapon/twohanded/rocketsledge/proc/get_fuel()
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:389]: no getters.

File: code/game/objects/items/weapons/twohanded.dm [hunk 5]
```diff
@@ -338,9 +338,162 @@
...
+/obj/item/weapon/twohanded/rocketsledge/wield(mob/user)
+	. = ..()
+	if(get_fuel() < fuel_used)
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 25)
+		return
+	icon_state = "rocketsledge1"
+
+/obj/item/weapon/twohanded/rocketsledge/unwield(mob/user)
+	. = ..()
+	icon_state = "rocketsledge"
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:401]: there already is wielded icon_states for these things. You shouldnt have to manually set them here.

File: code/game/objects/items/weapons/twohanded.dm [hunk 6]
```diff
@@ -338,9 +338,162 @@
...
+
+/obj/item/weapon/twohanded/rocketsledge/afterattack(obj/target, mob/user , flag)
+
+	if(istype(target, /obj/structure/reagent_dispensers/fueltank) && get_dist(user,target) <= 1)
+		var/obj/structure/reagent_dispensers/fueltank/FT = target
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:406]: same as before

File: code/game/objects/items/weapons/twohanded.dm [hunk 7]
```diff
@@ -338,9 +338,162 @@
...
+	if(get_fuel() < fuel_used)
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 25)
+		return
+	icon_state = "rocketsledge1"
+
+/obj/item/weapon/twohanded/rocketsledge/unwield(mob/user)
+	. = ..()
+	icon_state = "rocketsledge"
+
+/obj/item/weapon/twohanded/rocketsledge/afterattack(obj/target, mob/user , flag)
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:403]:
```suggestion
/obj/item/weapon/twohanded/rocketsledge/afterattack(obj/target, mob/user, flag)
```

File: code/game/objects/items/weapons/twohanded.dm [hunk 8]
```diff
@@ -338,9 +338,162 @@
...
+			to_chat(user, span_warning("Out of fuel!"))
+			return..()
+
+		var/fuel_transfer_amount = min(FT.reagents.total_volume, (max_fuel - get_fuel()))
+		FT.reagents.remove_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		reagents.add_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		playsound(loc, 'sound/effects/refill.ogg', 25, 1, 3)
+		to_chat(user, span_notice("You refill [src] with fuel."))
+
+	else
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:417]: early return, no `else`

File: code/game/objects/items/weapons/twohanded.dm [hunk 9]
```diff
@@ -338,9 +338,162 @@
...
+		..()
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:418]: you want to do `. = ..()`

File: code/game/objects/items/weapons/twohanded.dm [hunk 10]
```diff
@@ -338,9 +338,162 @@
...
+	. = ..()
+	icon_state = "rocketsledge"
+
+/obj/item/weapon/twohanded/rocketsledge/afterattack(obj/target, mob/user , flag)
+
+	if(istype(target, /obj/structure/reagent_dispensers/fueltank) && get_dist(user,target) <= 1)
+		var/obj/structure/reagent_dispensers/fueltank/FT = target
+		if(FT.reagents.total_volume == 0)
+			to_chat(user, span_warning("Out of fuel!"))
+			return..()
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:409]:
```suggestion
			return ..()
```

File: code/game/objects/items/weapons/twohanded.dm [hunk 11]
```diff
@@ -338,9 +338,162 @@
...
+
+	else
+		..()
+
+/obj/item/weapon/twohanded/rocketsledge/AltClick(mob/user)
+	if(!can_interact(user))
+		return ..()
+	if(!ishuman(user))
+		return ..()
+	if(!(user.l_hand == src || user.r_hand == src))
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:425]: all these checks can be one.

File: code/game/objects/items/weapons/twohanded.dm [hunk 12]
```diff
@@ -338,9 +338,162 @@
...
+		return ..()
+	TOGGLE_BITFIELD(flags_item, NODROP)
+	if(CHECK_BITFIELD(flags_item, NODROP))
+		to_chat(user, span_warning("You tighten the grip of [src] around your hand!"))
+	else
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:430]: early return

File: code/game/objects/items/weapons/twohanded.dm [hunk 13]
```diff
@@ -338,9 +338,162 @@
...
+		weaken = 2
+		stagger = 2
+		knockback = 0
+		to_chat(user, span_warning("Selected mode: CRUSH"))
+	else
+		stun = 1
+		weaken = 1
+		stagger = 1
+		knockback = 1
+		to_chat(user, span_warning("Selected mode: YEET"))
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:446]: "YEET". haha. Probably should change it tho.

File: code/game/objects/items/weapons/twohanded.dm [hunk 14]
```diff
@@ -338,9 +338,162 @@
...
+	playsound(loc, 'sound/machines/switch.ogg', 25)
+
+/obj/item/weapon/twohanded/rocketsledge/attack(mob/living/carbon/M as mob, mob/living/carbon/user as mob)
+	if(src && !CHECK_BITFIELD(src.flags_item, WIELDED))
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:450]: no, `src.this` or `src.that`
just do `this` or `that`

File: code/game/objects/items/weapons/twohanded.dm [hunk 15]
```diff
@@ -338,9 +338,162 @@
...
+		var/atom/throw_target = get_edge_target_turf(M, get_dir(src, get_step_away(M, src)))
+		var/throw_distance = knockback * LERP(5 , 2, M.mob_size / MOB_SIZE_BIG)
+		M.throw_at(throw_target, throw_distance, 0.5 + (knockback * 0.5), spin = TRUE)
+
+	if(isxeno(M))
+		var/mob/living/carbon/xenomorph/xeno_victim = M
+		if(xeno_victim.fortify) //If we're fortified we don't give a shit about staggerstun.
+			return
+
+		if(xeno_victim.endure) //Endure allows us to ignore staggerstun.
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:485]: these two can be one check

**Review by hyper2snyper — CHANGES_REQUESTED** [review]
Also make sure Kuro approves.

File: code/game/objects/items/weapons/twohanded.dm (hunk 5 above)
Comment (Waselon) [reply code/game/objects/items/weapons/twohanded.dm:401]: its for changing sprite in-hand, it doesnt change automatically

Comment (hyper2snyper) [reply code/game/objects/items/weapons/twohanded.dm]: thats what i mean. you are using wrong sprite names.

File: code/game/objects/items/weapons/twohanded.dm [hunk 16]
```diff
@@ -338,9 +338,161 @@
...
+
+/obj/item/weapon/twohanded/rocketsledge/dropped(mob/user)
+	. = ..()
+	toggle_item_bump_attack(user, FALSE)
+
+/obj/item/weapon/twohanded/rocketsledge/examine(mob/user)
+	..()
+	to_chat(user, "It contains [current_fuel()]/[max_fuel] units of fuel!")
+
+/obj/item/weapon/twohanded/rocketsledge/proc/current_fuel()
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:397]: no getters. just draw `reagents.get_reagent_amount(/datum/reagent/fuel)` everytime you need it.

File: code/game/objects/items/weapons/twohanded.dm [hunk 17]
```diff
@@ -338,9 +338,161 @@
...
+		return
+	icon_state = "rocketsledge1" //update sprite to "active" when we wield the weapon
+
+/obj/item/weapon/twohanded/rocketsledge/unwield(mob/user)
+	. = ..()
+	icon_state = "rocketsledge" //update sprite to "inactive" when we unwield the weapon
+
+/obj/item/weapon/twohanded/rocketsledge/afterattack(obj/target, mob/user, flag)
+
+	if(istype(target, /obj/structure/reagent_dispensers/fueltank) && get_dist(user,target) <= 1)
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:413]: early return

File: code/game/objects/items/weapons/twohanded.dm [hunk 18]
```diff
@@ -338,9 +338,161 @@
...
+		to_chat(user, span_notice("You refill [src] with fuel."))
+
+	. = ..()
+
+/obj/item/weapon/twohanded/rocketsledge/AltClick(mob/user)
+	if(!can_interact(user) || !ishuman(user) || !(user.l_hand == src || user.r_hand == src))
+		return ..()
+	TOGGLE_BITFIELD(flags_item, NODROP)
+	if(CHECK_BITFIELD(flags_item, NODROP))
+		to_chat(user, span_warning("You tighten the grip of [src] around your hand!"))
```
Comment (hyper2snyper) [inline code/game/objects/items/weapons/twohanded.dm:432]:
```suggestion
		to_chat(user, span_warning("You tighten the grip of [src] around your hand!"))
		return
```

**Review by hyper2snyper — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/twohanded.dm [hunk 19]
```diff
@@ -338,9 +338,161 @@
...
+/obj/item/weapon/twohanded/rocketsledge/wield(mob/user)
+	. = ..()
+	if(current_fuel() < fuel_used)
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 25)
+		return
+	icon_state = "rocketsledge1" //update sprite to "active" when we wield the weapon
+
+/obj/item/weapon/twohanded/rocketsledge/unwield(mob/user)
+	. = ..()
+	icon_state = "rocketsledge" //update sprite to "inactive" when we unwield the weapon
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:409]: all icon_state change must be done in update_icon_state()

File: code/game/objects/items/weapons/twohanded.dm [hunk 20]
```diff
@@ -338,9 +338,161 @@
...
+	else
+		stun = 1
+		weaken = 1
+		stagger = 1
+		knockback = 1
+		balloon_alert(user, "Selected mode: KNOCKBACK.")
+	playsound(loc, 'sound/machines/switch.ogg', 25)
+
+/obj/item/weapon/twohanded/rocketsledge/attack(mob/living/carbon/M as mob, mob/living/carbon/user as mob)
+	if(src && !CHECK_BITFIELD(flags_item, WIELDED))
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:452]: if(src)? what?

File: code/game/objects/items/weapons/twohanded.dm [hunk 21]
```diff
@@ -338,9 +338,161 @@
...
+		balloon_alert(user, "Selected mode: CRUSH.")
+	else
+		stun = 1
+		weaken = 1
+		stagger = 1
+		knockback = 1
+		balloon_alert(user, "Selected mode: KNOCKBACK.")
+	playsound(loc, 'sound/machines/switch.ogg', 25)
+
+/obj/item/weapon/twohanded/rocketsledge/attack(mob/living/carbon/M as mob, mob/living/carbon/user as mob)
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:451]: no as

File: code/game/objects/items/weapons/twohanded.dm [hunk 22]
```diff
@@ -338,9 +338,161 @@
...
+	attack_speed = 20
+
+	///amount of fuel stored inside
+	var/max_fuel = 50
+	///amount of fuel used per hit
+	var/fuel_used = 5
+	///additional damage when weapon is active
+	var/additional_damage = 50
+	///armor-piercing damage that ignores melee armor
+	var/ap_damage = 25
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:371]: this should be inside a state or mode datum

File: code/game/objects/items/weapons/twohanded.dm [hunk 23]
```diff
@@ -338,9 +338,161 @@
...
+			to_chat(user, span_warning("Out of fuel!"))
+			return ..()
+
+		var/fuel_transfer_amount = min(RS.reagents.total_volume, (max_fuel - current_fuel()))
+		RS.reagents.remove_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		reagents.add_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		playsound(loc, 'sound/effects/refill.ogg', 25, 1, 3)
+		to_chat(user, span_notice("You refill [src] with fuel."))
+
+	. = ..()
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:425]: return ..()

File: code/game/objects/items/weapons/twohanded.dm [hunk 24]
```diff
@@ -338,9 +338,161 @@
...
+		stagger = 2
+		knockback = 0
+		balloon_alert(user, "Selected mode: CRUSH.")
+	else
+		stun = 1
+		weaken = 1
+		stagger = 1
+		knockback = 1
+		balloon_alert(user, "Selected mode: KNOCKBACK.")
+	playsound(loc, 'sound/machines/switch.ogg', 25)
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:449]: early return

File: code/game/objects/items/weapons/twohanded.dm [hunk 25]
```diff
@@ -338,9 +338,161 @@
...
+		return ..()
+
+	M.apply_damage(max(0, additional_damage - additional_damage*M.hard_armor.getRating("melee")), BRUTE, user.zone_selected, M.get_soft_armor("melee", user.zone_selected))
+	M.apply_damage(ap_damage, BRUTE)
+	M.visible_message(span_danger("[user]'s rocket sledge hits [M.name], smashing them!"), span_userdanger("You [user]'s rocket sledge smashes you!"))
+
+	if(current_fuel() < fuel_used * 2)
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 50)
+		to_chat(user, span_warning("\The [src] shuts off, using last bits of fuel!"))
+		icon_state = "rocketsledge"
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:470]: icon_state change no

**Review by BraveMole — CHANGES_REQUESTED** [review]
125 damage PLUS ap is completly crazy

**Comment by Tristan63** [comment]
image

File: code/game/objects/items/storage/backpack.dm [hunk 26]
```diff
@@ -706,6 +706,18 @@
...
+		var/obj/item/weapon/twohanded/rocketsledge/RS = I
+		if(RS.reagents.get_reagent_amount(/datum/reagent/fuel) == RS.max_fuel || !reagents.total_volume)
+			return ..()
+
+		var/fuel_transfer_amount = min(reagents.total_volume, (RS.max_fuel - RS.reagents.get_reagent_amount(/datum/reagent/fuel)))
+		reagents.remove_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		RS.reagents.add_reagent(/datum/reagent/fuel, fuel_transfer_amount)
+		playsound(loc, 'sound/effects/refill.ogg', 25, 1, 3)
+		to_chat(user, span_notice("You refill [RS] with fuel."))
+		RS.update_icon_state()
```
Comment (BraveMole) [inline code/game/objects/items/storage/backpack.dm:719]: never call directly, call update_icon() (which in turns calls update_icon_state, with some signals and overlay changes as well)

File: code/game/objects/items/weapons/twohanded.dm [hunk 27]
```diff
@@ -338,9 +338,164 @@
...
+/obj/item/weapon/twohanded/rocketsledge/examine(mob/user)
+	..()
+	to_chat(user, "It contains [reagents.get_reagent_amount(/datum/reagent/fuel)]/[max_fuel] units of fuel!")
+
+/obj/item/weapon/twohanded/rocketsledge/wield(mob/user)
+	. = ..()
+	if((reagents.get_reagent_amount(/datum/reagent/fuel) < fuel_used))
+		playsound(loc, 'sound/items/weldingtool_off.ogg', 25)
+		return
+	update_icon_state()
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:400]: ditto

File: code/game/objects/items/weapons/twohanded.dm [hunk 28]
```diff
@@ -338,9 +338,164 @@
...
+	///additional damage when weapon is active
+	var/additional_damage = 75
+	///stun value
+	var/stun = 1
+	///weaken value
+	var/weaken = 2
+	///stagger value
+	var/stagger = 2
+	///knockback value
+	var/knockback = 0
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:377]: this would be far clearer in a datum

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/twohanded.dm [hunk 29]
```diff
@@ -338,9 +338,167 @@
...
+	update_icon()
+
+/obj/item/weapon/twohanded/rocketsledge/update_icon_state()
+	if ((reagents.get_reagent_amount(/datum/reagent/fuel) > fuel_used) && (CHECK_BITFIELD(flags_item, WIELDED)))
+		icon_state = "rocketsledge_w"
+	else
+		icon_state = "rocketsledge"
+
+/obj/item/weapon/twohanded/rocketsledge/update_icon()
+	update_icon_state()
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:413]: you should not need that. 
image
not sure why this exist, but this is bad. Check if you can safely remove

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/twohanded.dm [hunk 30]
```diff
@@ -338,9 +334,167 @@
...
+	. = ..()
+	update_icon()
+
+/obj/item/weapon/twohanded/rocketsledge/update_icon_state()
+	if ((reagents.get_reagent_amount(/datum/reagent/fuel) > fuel_used) && (CHECK_BITFIELD(flags_item, WIELDED)))
+		icon_state = "rocketsledge_w"
+	else
+		icon_state = "rocketsledge"
+
+/obj/item/weapon/twohanded/rocketsledge/update_icon()
```
Comment (BraveMole) [inline code/game/objects/items/weapons/twohanded.dm:408]: Remove it

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Review by Kuro020 — APPROVED** [review]

**Comment by Pariah919** [comment]
```
Added rocket sledge. Power weapon that can be optained through requisitions for 50 points.
has 2 modes (works only when fueled):

    stun = 1, weaken = 2, stagger = 2, knockback = 0;
    stun = 1, weaken = 1, stagger = 1, knockback = 1;

75 damage without fuel;
150 damage with fuel (active);
attack speed = 20 (2 sec delay);```
this has bump???
this thing is utterly broken for 50 points, remove bump from this and remove the stun or make it stutter stun, not sure about the stagger but this thing seems just utterly insane

**Pull request merged**